### PR TITLE
Secrets manager introduced to commands

### PR DESCRIPTION
**Pull request by Nemanja0x**
# Description
Commands stake, unstake, withdraw, validator info use secret manager to provide access to cloud secret storage or local disc.
Naming of the flags are standardized for all commands: 
        DataPathFlag = "data-dir"  - for local storage
	ConfigFlag   = "config"       - cloud config

# Changes include

- [ ] Bugfix (non-breaking change that solves an issue)
- [ ] Hotfix (change that solves an urgent issue, and requires immediate attention)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (change that is not backwards-compatible and/or changes current functionality)

# Checklist

- [x] I have assigned this PR to myself
- [ ] I have added at least 1 reviewer
- [ ] I have added the relevant labels
- [ ] I have updated the official documentation
- [ ] I have added sufficient documentation in code

## Testing

- [x] I have tested this code with the official test suite
- [ ] I have tested this code manually

### CODE REVIEW

File: command/sidechain/validators/validator_info.go [hunk 1]
```diff
@@ -47,7 +56,13 @@ func runCommand(cmd *cobra.Command, _ []string) error {
 	outputter := command.InitializeOutputter(cmd)
 	defer outputter.WriteOutput()
 
-	validatorAccount, err := sidechainHelper.GetAccountFromDir(params.accountDir)
+	// get secret manager and allow reading from local directory (true at the end)
```
Comment (igorcrevar) [inline command/sidechain/validators/validator_info.go]: Maybe this should be inside a helper function. The same boilerplate code is being used more than a few times.:
```
// get secret manager and allow reading from local directory (true at the end)
	secretsManager, err := polybftsecrets.GetSecretsManager(params.accountDir, params.configPath, true)
	if err != nil {
		return err
	}

	validatorAccount, err := wallet.NewAccountFromSecret(secretsManager)
	```

Comment (Nemanja0x) [reply command/sidechain/validators/validator_info.go]: fixed: https://github.com/0xPolygon/polygon-edge/pull/1251/commits/755531f6d8693bcd0e4717457c3cbaa2160475ab

**Review by igorcrevar — APPROVED** [review]

File: command/polybftsecrets/utils.go [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+	ConfigFlagDesc   = "the path to the SecretsManager config file, if omitted, the local FS secrets manager is used"
+)
+
+// common errors for all polybft commands
+var (
+	ErrInvalidNum                     = fmt.Errorf("num flag value should be between 1 and %d", maxInitNum)
+	ErrInvalidConfig                  = errors.New("invalid secrets configuration")
+	ErrInvalidParams                  = errors.New("no config file or data directory passed in")
+	ErrUnsupportedType                = errors.New("unsupported secrets manager")
+	ErrSecureLocalStoreNotImplemented = errors.New(
```
Comment (Stefan-Ethernal) [inline command/polybftsecrets/utils.go:26]: Nitpick: You could have made all errors, except `ErrInvalidParams` local (as they are used only in `polybftsecrets` package).

File: command/sidechain/withdraw/withdraw.go [hunk 3]
```diff
@@ -37,9 +38,16 @@ func GetCommand() *cobra.Command {
...
-		"the directory path where validator key is stored",
+		polybftsecrets.DataPathFlagDesc,
+	)
+
+	cmd.Flags().StringVar(
+		&params.configPath,
+		polybftsecrets.ConfigFlag,
+		"",
+		polybftsecrets.ConfigFlagDesc,
 	)
```
Comment (Stefan-Ethernal) [inline command/sidechain/withdraw/withdraw.go:51]: Make `DataPathFlag` and `ConfigFlag` mutually exclusive?

File: command/sidechain/validators/validator_info.go [hunk 4]
```diff
@@ -31,9 +32,16 @@ func GetCommand() *cobra.Command {
...
-		"the directory path where validator key is stored",
+		polybftsecrets.DataPathFlagDesc,
+	)
+
+	cmd.Flags().StringVar(
+		&params.configPath,
+		polybftsecrets.ConfigFlag,
+		"",
+		polybftsecrets.ConfigFlagDesc,
 	)
```
Comment (Stefan-Ethernal) [inline command/sidechain/validators/validator_info.go:45]: `DataPath` and `ConfigFlag` mutually exclusive?

File: command/sidechain/helper.go [hunk 5]
```diff
@@ -33,15 +32,32 @@ func CheckIfDirectoryExist(dir string) error {
 	return nil
 }
 
-func GetAccountFromDir(dir string) (*wallet.Account, error) {
-	secretsManager, err := secretsHelper.SetupLocalSecretsManager(dir)
+func ValidateSecretFlags(dataDir, config string) error {
+	if config == "" {
```
Comment (Stefan-Ethernal) [inline command/sidechain/helper.go:36]: Should you validate provided `config` value (at least you could check if the directory exists)?

File: command/sidechain/helper.go [hunk 6]
```diff
@@ -33,15 +32,32 @@ func CheckIfDirectoryExist(dir string) error {
...
+		} else {
+			return CheckIfDirectoryExist(dataDir)
+		}
+	}
+
+	return nil
+}
+
+func GetAccount(dataDir, config string) (*wallet.Account, error) {
+	// get secret manager and allow reading from local directory (true at the end)
```
Comment (Stefan-Ethernal) [inline command/sidechain/helper.go]:
```suggestion
	// resolve secrets manager instance and allow usage of insecure local secrets manager
```

File: command/sidechain/helper.go [hunk 7]
```diff
@@ -33,15 +32,32 @@ func CheckIfDirectoryExist(dir string) error {
...
 	if err != nil {
 		return nil, err
 	}
 
 	return wallet.NewAccountFromSecret(secretsManager)
 }
 
+func GetAccountFromDir(dir string) (*wallet.Account, error) {
+	return GetAccount(dir, "")
+}
```
Comment (Stefan-Ethernal) [inline command/sidechain/helper.go:59]: I don't see many benefits of this wrapper function, but it's your call whether you want to keep it.

File: command/polybftsecrets/utils.go [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+		"use a secrets backend, or supply an --insecure flag " +
+			"to store the private keys locally on the filesystem, " +
+			"avoid doing so in production")
+)
+
+func GetSecretsManager(dataPath, configPath string, insecureLocalStore bool) (secrets.SecretsManager, error) {
```
Comment (Stefan-Ethernal) [inline command/polybftsecrets/utils.go:34]: Consider adding a comment.

**Review by Stefan-Ethernal — COMMENTED** [review]
Generally looks good.

Points to consider addressing:
- What should we do with `command/rootchain/emit/emit.go` and `command/rootchain/initcontracts/init_contracts.go`? Those two commands are expecting hex-encoded private key to be provided via the CLI flag, but not sure if is it ok to leave it like that.
- Also just mark `DataPathFlag` and `ConfigFlag` mutually exclusive in each command it accommodates it.

File: command/polybftsecrets/utils.go (hunk 2 above)
Comment (Nemanja0x) [reply command/polybftsecrets/utils.go:26]: There  is another task to split register command. For those purposes I'll need these errors. 

`What should we do with command/rootchain/emit/emit.go and command/rootchain/initcontracts/init_contracts.go? Those two commands are expecting hex-encoded private key to be provided via the CLI flag, but not sure if is it ok to leave it like that.`  - this should be addressed via another task.

`Also just mark DataPathFlag and ConfigFlag mutually exclusive in each command it accommodates it.`  - will do

File: command/sidechain/helper.go (hunk 7 above)
Comment (Nemanja0x) [reply command/sidechain/helper.go:59]: Will leave it for tests.

**Review by b00ris — APPROVED** [review]

File: command/sidechain/helper.go (hunk 5 above)
Comment (Nemanja0x) [reply command/sidechain/helper.go:36]: That is done while creating secrets manager. Proper message will be returned.

File: command/polybftsecrets/utils.go (hunk 8 above)
Comment (Nemanja0x) [reply command/polybftsecrets/utils.go:34]: done: https://github.com/0xPolygon/polygon-edge/pull/1251/commits/5f2f7453acce35cc3e7352c75442459c60908a31

File: command/sidechain/helper.go (hunk 6 above)
Comment (Nemanja0x) [reply command/sidechain/helper.go]: done: https://github.com/0xPolygon/polygon-edge/pull/1251/commits/5f2f7453acce35cc3e7352c75442459c60908a31

File: command/sidechain/validators/validator_info.go (hunk 4 above)
Comment (Nemanja0x) [reply command/sidechain/validators/validator_info.go:45]: done: https://github.com/0xPolygon/polygon-edge/pull/1251/commits/5f2f7453acce35cc3e7352c75442459c60908a31

File: command/sidechain/withdraw/withdraw.go (hunk 3 above)
Comment (Nemanja0x) [reply command/sidechain/withdraw/withdraw.go:51]: done: https://github.com/0xPolygon/polygon-edge/pull/1251/commits/5f2f7453acce35cc3e7352c75442459c60908a31

File: command/polybftsecrets/utils.go (hunk 2 above)
Comment (Nemanja0x) [reply command/polybftsecrets/utils.go:26]: done: https://github.com/0xPolygon/polygon-edge/pull/1251/commits/5f2f7453acce35cc3e7352c75442459c60908a31

**Review by goran-ethernal — APPROVED** [review]

File: command/sidechain/helper.go (hunk 5 above)
Comment (Stefan-Ethernal) [reply command/sidechain/helper.go:36]: As far as I can see, it is going to try and read a config file here:
https://github.com/0xPolygon/polygon-edge/blob/develop/secrets/config.go#L29-L33. Maybe we could add there explicit check if file doesn't exist and return a proper error message (not sure if `os.ReadFile` is going to provide a meaningful error message)?

**Review by Stefan-Ethernal — APPROVED** [review]

**Review by vcastellm — COMMENTED** [review]
Late review but shared flags between subcommands can be defined in the top level command if I'm not wrong.

**Pull request merged**